### PR TITLE
added isolation commands to cli (GSI-689)

### PR DESCRIPTION
**Pull request by KerstenBreuer**
```
- increased tests for isolate functions
- make cli tests more behavior driven
```

Depends on #28.

### CODE REVIEW

File: src/schemapack/_internals/cli/main.py [hunk 1]
```diff
@@ -148,3 +151,122 @@ def condense_schemapack(
...
+    *,
+    schemapack: Annotated[
+        Path,
+        typer.Option(
+            "--schemapack",
+            "-s",
+            help=(
+                "Provide the path to a schemapack that describes the structure of"
+                " provided input datapack (not the rooted datapack output by this"
+                " command)."
```
Comment (lkuchenb) [inline src/schemapack/_internals/cli/main.py:238]:
```suggestion
                "The path to a schemapack that describes the structure of the"
                " provided input datapack (not the rooted datapack output by this"
                " command)."
```

File: src/schemapack/_internals/dump.py [hunk 2]
```diff
@@ -115,14 +91,23 @@ def dumps_schemapack(
...
-    schemapack_dict = schemapack_to_dict(schemapack)
+    return dumps_model(schemapack, yaml_format=yaml_format)
+
 
-    if yaml_format:
-        with StringIO() as buffer:
-            yaml.dump(schemapack_dict, buffer)
-            return buffer.getvalue().strip()
+def dumps_datapack(
+    datapack: BaseModel,
```
Comment (lkuchenb) [inline src/schemapack/_internals/dump.py:98]:
```suggestion
    datapack: DataPack,
```

File: src/schemapack/_internals/dump.py [hunk 3]
```diff
@@ -16,18 +16,17 @@
...
 
-import json
-from io import StringIO
 from pathlib import Path
 from typing import Any
 
-import ruamel.yaml
+from pydantic import BaseModel
 
 from schemapack._internals.spec.schemapack import SchemaPack
```
Comment (lkuchenb) [inline src/schemapack/_internals/dump.py:24]:
```suggestion
from schemapack._internals.spec.datapack import DataPack
from schemapack._internals.spec.schemapack import SchemaPack
```

File: src/schemapack/_internals/exceptions.py [hunk 4]
```diff
@@ -204,3 +204,44 @@ class CircularRelationError(BaseError, ValueError):
     """Raised when a circular relation between resources is detected, but the requested
     operation cannot be performed on datapacks with circular relations.
     """
+
+
+class ClassNotFoundError(BaseError, KeyError):
+    """Raised when a class was not found in a schemapack or datapack."""
+
+    def __init__(self, *, class_name: str, in_schemapack: bool):
```
Comment (lkuchenb) [inline src/schemapack/_internals/exceptions.py:212]: Maybe a `ModelType` enum would be a bit more self-explanatory than the `in_schemapack` bool.

File: src/schemapack/_internals/exceptions.py [hunk 5]
```diff
@@ -204,3 +204,44 @@ class CircularRelationError(BaseError, ValueError):
...
+
+        Args:
+            class_name:
+                The name of the class of the resource that was not found.
+            resource_id:
+                The ID of the resource that was not found.
+        """
+        message = (
+            f"A resource of class '{class_name}' with id '{resource_id}' not found in"
+            + " the provided datapack."
```
Comment (lkuchenb) [inline src/schemapack/_internals/exceptions.py:243]:
```suggestion
            f"Resource of class '{class_name}' with id '{resource_id}' not found in"
            + " the provided datapack."
```

File: src/schemapack/_internals/utils.py [hunk 6]
```diff
@@ -93,3 +98,47 @@ def transient_directory_change(path: Path):
...
+
+    if yaml_format:
+        with StringIO() as buffer:
+            yaml.dump(model_dict, buffer)
+            return buffer.getvalue().strip()
+
+    return json.dumps(model_dict, indent=2)
+
+
+def write_dict(schemapack_dict: dict, *, path: Path, yaml_format: bool) -> None:
```
Comment (lkuchenb) [inline src/schemapack/_internals/utils.py:137]: The function name suggests this is a generic function, yet the argument `schemapack_dict` suggests it's not

File: tests/test_isolate.py [hunk 7]
```diff
@@ -93,3 +100,69 @@ def test_isolate(
...
+    schemapack = load_schemapack(schemapack_path)
+    datapack = load_datapack(datapack_path)
+
+    with pytest.raises(exceptions.ClassNotFoundError):
+        isolate(
+            class_name="NonExistingClass",
+            resource_id="example_dataset_1",
+            schemapack=schemapack,
+            datapack=datapack,
+        )
```
Comment (lkuchenb) [inline tests/test_isolate.py:118]:
```suggestion
    with pytest.raises(exceptions.ClassNotFoundError) as error:
        isolate(
            class_name="NonExistingClass",
            resource_id="example_dataset_1",
            schemapack=schemapack,
            datapack=datapack,
        )
    assert "schemapack" in str(error)
```

Comment (lkuchenb) [reply tests/test_isolate.py:118]: Maybe add these types of assertions to the `ClassNotFoundError` tests?

**Review by lkuchenb — CHANGES_REQUESTED** [review]
LGTM, some minor suggestions below.

File: src/schemapack/_internals/exceptions.py (hunk 4 above)
Comment (KerstenBreuer) [reply src/schemapack/_internals/exceptions.py:212]: I agree, thanks

File: src/schemapack/_internals/utils.py (hunk 6 above)
Comment (KerstenBreuer) [reply src/schemapack/_internals/utils.py]: Right, I  renamed it generically.

**Review by lkuchenb — DISMISSED** [review]

**Review by lkuchenb — APPROVED** [review]

**Pull request merged**